ligo/haskell/src/Ligo/Typescript.hs

```diff
@@ -0,0 +1,380 @@
+  build (TsInterface fns) =  Fmt.unlinesF ["{", Fmt.indentF 2 $ Fmt.unlinesF $ mkPair <$> fns, "}"]
+    where
+      mkPair :: (Fieldname, (Typename, IsOptional)) -> Text
+      mkPair (fn, (tn, isOptional)) =
+        case isOptional of
+          True -> fn <> "?:" <> tn
+          False -> fn <> ":" <> tn
+  build (TsAlias t) = build t
+
+-- Implements the conversion to source code.
+instance Buildable TsDecl where
+  build (TsType typename tdef) = case tdef of
+    TsSum _ -> Fmt.unwordsF ["export type", build typename, "=", build tdef]
+    TsTuple _ -> Fmt.unwordsF ["export type", build typename, "=", build tdef]
+    TsInterface _ -> Fmt.unwordsF ["export interface", build typename, build tdef]
+    TsAlias _ -> Fmt.unwordsF ["export type", build typename, "=", build tdef]
+
+-- Write down Typescript modules to represents types for the contract parameter 'cp'
+generateTs :: forall cp m. (MonadIO m, ParameterDeclaresEntrypoints cp) => FilePath -> m FilePath
+generateTs fp = do
+  let epModules = generateEpModule (pnNotes $ parameterEntrypointsToNotes @cp)
+  mapM_ (writeModule fp) epModules
+  let imports = (\ep -> "import * as "
+        <> (tsmEpName ep)
+        <> " from  './" <> (tsmEpName ep) <> "';") <$> epModules
+  let parameterName = "Parameter"
+  let parameterModule = TsModule
+        parameterName (toText <$> imports)
+          [TsType parameterName $
+            TsSum $ (\x -> let n = tsmEpName x in (n <> "." <> n)) <$> epModules]
+  writeModule fp parameterModule
+
+-- Convert a 'Notes t' structure to a list of modules containing types
+-- that wraps its various entrypoints. Each entrypoint is wrapped into
+-- its own modules.
+generateEpModule :: Notes t -> [TsModule]
+generateEpModule notes = let
+  eps = lookupEntryPointsAndTypes notes
+  epNameCaps x = T.cons (toUpper $ T.head x) (T.tail x)
+  mkModule Entrypoint {..} = let
+    capsdEp = epNameCaps epName
+    in TsModule capsdEp [] $ mkTypeFor capsdEp epType
+  in mkModule <$> eps
+
+-- Given a parent directory and a TsModule, write down the module as a
+-- Typescript source file in the parent directory.
+writeModule :: MonadIO m => FilePath -> TsModule -> m FilePath
+writeModule fp (TsModule name imports decls) = do
+  let filename = (toString name) <.> "ts"
+  let declLines = build <$> decls
+  let importsLines = build <$> imports
+  writeFile (fp </> filename) (Fmt.pretty $ Fmt.unlinesF (importsLines <> declLines))
+  pure filename
+
+-- Given a typename, and a type, generated Typescript declarations for this
+-- type and all of its fields. Pairs are coverted into Tuples or Interfaces.
+-- If there is atleast one named field in the type, it is converted into
+-- interface.  Or else into a Typescript tuple. 'Or' structures are converted
+-- into Sum types after flattening their structure.  Rest of the types are
+-- converted into corresponding Typescript type after creating type aliased for
+-- their inner fields.
```

Comment:
I'll make a suggestion, as I think this is easier to read:

```haskell
-- Given a typename, and a type, generates Typescript declarations for this
-- type and all of its fields. If there is at least one named field in the type, it is
-- converted into an interface, otherwise into a TypeScript tuple. 'Or' structures are
-- converted into Union types after flattening their structure. The rest of the types are
-- converted into the corresponding TypeScript types after creating type aliases for
-- their inner fields.
```

Reply:
Have replaced it with your suggestion which indeed reads better.